### PR TITLE
Non-deflected and non-fully-penetrating attacks deal extra blunt damage.

### PR DESCRIPTION
**Pull request by AluminumAlman**
## Changes

- Changes to sharp damage damage amount mitigation code: sharp attacks that aren't deflected, but which's damage amount was reduced by armor, deal extra blunt damage, which's damage amount and armor penetration is based off damage amount/armor penetration that was blocked by armor;
- Modified the CE: New armor system learning helper and sharp armor stat description.

## Reasoning

- With the current system, projectiles which do not penetrate armor may be more damaging than projectiles which barely penetrate armor, for example AP bullets which penetrate a devilstrand parka by a hair deal less damage than FMJ/HP bullets that were deflected by a devilstrand parka. This change is supposed to fix that issue;
- Done to account for the changes this PR does to the armor system.

## Testing

- [x] Compiles without warnings;
- [x] Game runs without errors;
- [ ] (For compatibility patches) ...with and without patched mod loaded;
- [ ] (Dedicatedly) playtested a colony;
- Playing with the change since 2021-04-28 (YYYY-MM-DD).

## Assembly
[CombatExtended.zip](https://github.com/CombatExtended-Continued/CombatExtended/files/6610225/CombatExtended.zip)

### CODE REVIEW

**Comment by Breadbox2k19** [comment]
Good change in general, but it might prompt a review of blunt penetration of projectiles and melee attack in general, the current blunt penetration of bullets is simply not balanced in anyway and this change will make thing alot worse.
An example being intermediate rounds vs full power rounds, with the latter already being superior than the former even without the change.

I recommend lowering the blunt penetration of projectiles in general, using realistic values would simply make most non full caliber rifle rounds obsolete.

**Review by Breadbox2k19 — APPROVED** [review]
It is better for the rounds to be unbalanced than to live with a broken armour system.

**Comment by Breadbox2k19** [comment]
I've only playtested it alittle, so it would be good if someone else give a second opinion before merging.

File: Defs/Tutor/Concepts_NotedOpportunistic.xml [hunk 1]
```diff
@@ -122,7 +122,7 @@
...
...rate weaponless attack.\n\nIf a sharp attack's damage is not affected by any armor, then it deals its original damage.\n\nIf a blunt attack has less blunt penetration than the amount of blunt protection the armor offers, then the attack is completely blocked (e.g: a pawn with 45 MPa worth of blunt armor that is hit with an attack that has 20 MPa blunt penetration takes no blunt damage from the blunt attack).\n\nIf a blunt attack has more blunt penetration than the amount of blunt protection the armor offers, then the damage amount is reduced by how much blunt penetration was blocked (e.g: a pawn with 12 MPa worth of blunt armor that is hit with an attack that has 20 MPa blunt penetration only receives 40% of the attack's original damage).\n\nIf a blunt attack's damage is not affected by any armor, then it deals its original damage.\n\nEnvironmental damage such as fire/electricity has no penetration and is simply reduced by the % of armor against that type.\n\nArmor made from hard materials like steel or composite still take minimal damage from blocked attacks, however armor made from soft materials like cloth or devilstrand take no damage from deflected or blunt attacks.</helpText>
```
Comment (LockdownX7) [inline Defs/Tutor/Concepts_NotedOpportunistic.xml:125]: Learning helper content should be concise, for the following reasons:
1. There's no scrolling support, so any part of the learning helper that goes off-screen is inacessible. You've put so much text in here that it's missing four entire paragraphs when the game is running at 1920x1080 with 1.0x UI scale.
2. A 3,000-character block of text is hard to take for a new player just trying to learn the CE basics. If you take the previous text and just edit it slightly to mention that a projectile progressively converts part of its sharp damage to blunt as a target's armor gets closer to the projectile's armor penetration, that should be informative enough to get the ball rolling. If later on they want to dive into the specifics, they can check the armor info panels, which is where you should provide an in-depth explanation of sharp/blunt armor penetration and deflection. You can edit them at `CombatExtended\Patches\Core\Stats\Stats.xml`

File: Source/CombatExtended/CombatExtended/ArmorUtilityCE.cs [hunk 2]
```diff
@@ -347,6 +357,61 @@ private static DamageInfo GetDeflectDamageInfo(DamageInfo dinfo, BodyPartRecord
...
 
+        /// <summary>
+        /// Creates a new DamageInfo for blunt damage based off the blocked amount of damage and penetration of the supplied DamageInfo.
+        /// </summary>
+        /// <param name="dinfo">The damage info the additional blunt damage will be based off</param>
+        /// <param name="hitPart">The originally hit part</param>
+        /// <param name="dmgAmount">The amount of damage blocked</param>
+        /// <param name="penAmount">The amount of penetration blocked</param>
+        /// <returns>New DamageInfo based off dinfo with damage, penetration, Def and forceHitPart adjusted</returns>>
+        private static DamageInfo GetPartialPenetrationBluntDamageInfo(DamageInfo dinfo, BodyPartRecord hitPart, float dmgAmount, float penAmount) // Not referencing the float values directly since we don't want to update them
```
Comment (LockdownX7) [inline Source/CombatExtended/CombatExtended/ArmorUtilityCE.cs:368]: This and GetDeflectDamageInfo have a lot of duplicate code between them, they should be merged into a single function that handles both use cases. I included an example of a modified GetDeflectDamageInfo function that incorporates your changes, but it's not tested at all, so if you integrate it wholesale, please make sure it works as expected:
https://gist.github.com/LockdownX7/45b299ba14c23b608a80b22b618f3a86

**Review by LockdownX7 — CHANGES_REQUESTED** [review]

**Comment by AluminumAlman** [comment]
Assembly with the most recent changes: [CombatExtended.zip](https://github.com/CombatExtended-Continued/CombatExtended/files/6609543/CombatExtended.zip)

**Comment by AluminumAlman** [comment]
Don't merge for now, I'm doing more testing.

File: Defs/Tutor/Concepts_NotedOpportunistic.xml (hunk 1 above)
Comment (AluminumAlman) [reply Defs/Tutor/Concepts_NotedOpportunistic.xml]: Changed the learning helper text, now it's only twice as big (as opposed to four times), at the same time edited the description of the sharp armor stat.

File: Source/CombatExtended/CombatExtended/ArmorUtilityCE.cs (hunk 2 above)
Comment (AluminumAlman) [reply Source/CombatExtended/CombatExtended/ArmorUtilityCE.cs]: Merged them both into one thanks to you, although I did modify it to work correctly.

**Review by LockdownX7 — APPROVED** [review]

**Review by MadaraUchiha — APPROVED** [review]
LGTM :+1:

**Review by N7Huntsman — APPROVED** [review]

**Pull request merged**